dynamic_instrumentation/src/instrument.rs

```diff
@@ -343,43 +343,110 @@ impl<'tcx> Visitor<'tcx> for CollectInstrumentationPoints<'_, 'tcx> {
-                            .copied()
-                            .filter(|_| base.projection.len() == place.projection.len() - 1)
-                    };
-                    self.loc(location, location, field_fn)
-                        .arg_var(place.local)
-                        .arg_index_of(field)
-                        .source(place)
-                        .dest_from(proj_dest)
-                        .add_to(self);
+        if !place.is_indirect() {
+            return;
+        }
+        if !context.is_use() {
+            return;
+        }
+
+        let copy_fn = self.hooks().find("ptr_copy");
+        let project_fn = self.hooks().find("ptr_project");
+        let load_fn = self.hooks().find("ptr_load");
+        let load_value_fn = self.hooks().find("load_value");
+        let store_fn = self.hooks().find("ptr_store");
+
+        // Instrument nested projections and derefs
+        let mut proj_iter = place.iter_projections().peekable();
+
+        // Skip over the initial projections of the local since we do
+        // not want to take its address; we only care about the ones
+        // past the first dereference since at that point we have an
+        // actual pointer.
+        while let Some((_, elem)) = proj_iter.peek() {
+            if matches!(elem, PlaceElem::Deref) {
+                break;
+            }
+            let _ = proj_iter.next();
+        }
+
+        while let Some((deref1_base, PlaceElem::Deref)) = proj_iter.next() {
+            // Find the next Deref or end of projections
+            let (deref2_base, have_deref2) = loop {
+                match proj_iter.peek() {
+                    Some((base, PlaceElem::Deref)) => break (base.clone(), true),
+                    // Reached the end, we can use the full place
+                    None => break (place.as_ref(), false),
+                    _ => {
+                        let _ = proj_iter.next();
+                    }
                 }
+            };
+
+            // We have some other elements between the two projections
+            if deref2_base.projection.len() - deref1_base.projection.len() > 1 {
+                self.loc(location, location, project_fn)
+                    .arg_var(deref1_base)
+                    .arg_addr_of(deref2_base.clone())
+                    .add_to(self);
+            }
+
+            use PlaceContext::*;
+            if let Some(ptr_fn) = match context {
+                // We are just loading the address for the next dereference
+                _ if have_deref2 => Some(load_fn),
+
+                NonMutatingUse(NonMutatingUseContext::Inspect) |
+                NonMutatingUse(NonMutatingUseContext::Copy) |
+                NonMutatingUse(NonMutatingUseContext::Move) => Some(load_fn),
+
+                MutatingUse(MutatingUseContext::Store) |
+                MutatingUse(MutatingUseContext::Call) |
+                MutatingUse(MutatingUseContext::AsmOutput) => Some(store_fn),
```

Comment:
In cases with 3+ derefs, is it correct to use `store_fn` for the derefs in the middle?  My usual intuition is that all `Deref` projections except the last are loads, and the last one depends on the context.  Given `*(*(*p).x).y`, my understanding is that the loop will run twice, once with `p` and `(*p).x` as `deref1_base` and `deref2_base`, and again with `(*p).x` and `(*(*p).x).y`.  But only the last of these is actually the target of a store - nothing stores to the intermediate `(*p).x`, even though it appears as `deref2_base` the first time around the loop.

Reply:
Line 397 is supposed to cover this case (`have_deref2` is essentially equivalent to "is not the last deref"). We should get here 3 times with the following `(deref1_base, deref2_base)` pairs:
* `p`, `(*p).x` with `have_deref2==true`, will always be a load because of 397
* `(*p).x)`, `(*(*p).x).y` with `have_deref2==true`, will always be a load because of 397
* `*(*p).x).y`, `*(*(*p).x).y` with `have_deref2==false` because of line 379, will be a store only if the place context needs it. `deref2_base` and `have_deref2` might be confusing here. There is no explicit dereference in the MIR, only the implicit one from reading/writing the place itself (e.g. `(*_p).x.y` is counted as an extra implicit dereference of the final pointer, so 2 in total).

Reply:
That makes sense, thanks